gaurun/worker.go

```diff
@@ -3,53 +3,96 @@ package gaurun
 import (
 	"github.com/RobotsAndPencils/buford/push"
 	"strings"
+	"sync/atomic"
 )
 
+var (
+	PusherCount int64
+)
+
+func init() {
+	PusherCount = 0
+}
+
 func StartPushWorkers(workerNum, queueNum int) {
 	QueueNotification = make(chan RequestGaurunNotification, queueNum)
 	for i := 0; i < workerNum; i++ {
 		go pushNotificationWorker()
 	}
 }
 
+func isExternalServerError(err error, platform int) bool {
+	switch platform {
+	case PlatFormIos:
+		if err == push.ErrIdleTimeout || err == push.ErrShutdown || err == push.ErrInternalServerError || err == push.ErrServiceUnavailable {
+			return true
+		}
+	case PlatFormAndroid:
+		if err.Error() == "Unavailable" || err.Error() == "InternalServerError" || strings.Contains(err.Error(), "Timeout") {
+			return true
+		}
+	default:
+		// not through
+	}
+	return false
+}
+
+func pushSync(pusher func(req RequestGaurunNotification) error, req RequestGaurunNotification, retryMax int) {
+Retry:
+	err := pusher(req)
+	if err != nil && req.Retry < retryMax && isExternalServerError(err, req.Platform) {
+		req.Retry++
+		goto Retry
+	}
+}
+
+func pushAsync(pusher func(req RequestGaurunNotification) error, req RequestGaurunNotification, retryMax int) {
+Retry:
+	err := pusher(req)
+	if err != nil && req.Retry < retryMax && isExternalServerError(err, req.Platform) {
```

Comment:
The notification is dropped silently if it fails retryMax times, it should be logged?